Deploying a trained model to a docker image

savagej opened the pull request.

Commit:
Updated code to save a json dump of model details

Commit:
Fix bugs for outputting file

Also fixed requirements for Dockerfile-trainer

michelav commented on Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/news_classifier.py:
```diff
@@ -395,7 +418,18 @@ def get_predictions(self, model, data_loader):
-    mlflow.pytorch.log_model(model, "bert-model", registered_model_name="BertModel", extra_files=["class_mapping.json", "bert_base_uncased_vocab.txt"])
+    mlflow.pytorch.log_model(model, "bert-model", registered_model_name=args.model_name, extra_files=["class_mapping.json", "bert_base_uncased_vocab.txt"])
+
+    if args.json_dump is not None:
+        from mlflow.entities.model_registry import ModelVersion
+
+        client = MlflowClient()
+        filter_string = f"run_id='{run_id}'"
+        registered_model: ModelVersion = client.search_model_versions(filter_string=filter_string)[0]
+        model_details = {k.strip("_"): v for k, v in registered_model.__dict__.items()}
```
In order to avoid this save and search, would this work out?

```suggestion
    if args.json_dump is not None:
        from mlflow.entities.model_registry import ModelVersion

        client = MlflowClient()
        model_version = mlflow.register_model(f"runs:/{run_id}/artifacts/bert-model", args.model_name)
        model_details = {k.strip("_"): v for k, v in model_version.__dict__.items()}
```

Check here: https://mlflow.org/docs/latest/model-registry.html#adding-an-mlflow-model-to-the-model-registry

savagej replied:
Perfect, this is much better, we can be much more explicit about registering model this way, and avoids the race conditions/searching

Commit:
Updates for transformers

Commit:
Get model version by registering model manually

Commit:
Merge remote-tracking branch 'origin/master' into train-build

# Conflicts:
#	Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-trainer
#	Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/MLproject
#	Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/requirements.txt

Commit:
Working train and build of torchserve image

Took the entrypoint from torchserve

michelav commented:
Well... that start-deploy-stop thing in _Dockerfile-torchserve_ was pretty awesome. Seems much better than creating the model deployment when running the container. :+1:

Commit:
Updated train-and-build to build training image and push built serve image

Changed json dump file to prevent old files being read. Also fixed typos in news_classifier.py and MLproject

savagej commented:
Able to train a model, build the deploy image and push to the registry: https://console.cloud.google.com/gcr/images/engineeringlab/GLOBAL/labs1-mlflow-torch?project=engineeringlab
Definitely open to suggestions for which parts of `train-and-build.sh` make sense in the GH action

michelav reviewed (approved):
It's working and all other tasks will need this to keep going. Better to merge to Master as soon as possible.
